pennylane_forest/qvm.py

```diff
@@ -165,80 +144,66 @@ def pre_measure(self):
 
         self.prog = prag + self.prog
 
-        qubits = sorted(self.prog.get_qubits())
+        qubits = sorted(self.wiring.values())
         ro = self.prog.declare("ro", "BIT", len(qubits))
         for i, q in enumerate(qubits):
             self.prog.inst(MEASURE(q, ro[i]))
 
         self.prog.wrap_in_numshots_loop(self.shots)
 
-        if "pyqvm" in self.qc.name:
-            bitstring_array = self.qc.run(self.prog)
-        else:
-            self.compiled = self.qc.compile(self.prog)
-            bitstring_array = self.qc.run(executable=self.compiled)
-
-        self.state = {}
-        for i, q in enumerate(qubits):
-            self.state[q] = bitstring_array[:, i]
-
-    def expval(self, observable, wires, par):
-        return np.mean(self.sample(observable, wires, par))
-
-    def var(self, observable, wires, par):
-        return np.var(self.sample(observable, wires, par))
-
-    def sample(self, observable, wires, par):
-        wires = [self.wiring[i] for i in wires]
-        n = self.shots
-
-        if observable == "Identity":
-            return np.ones([n])
-
-        if observable == "Hermitian":
-            Hkey = tuple(par[0].flatten().tolist())
-            eigvals = self._eigs[Hkey]["eigval"]
-            res = np.array([self.state[i] for i in wires]).T
-
-            samples = np.zeros([n])
-
-            for w, b in zip(eigvals, itertools.product([0, 1], repeat=len(wires))):
-                samples = np.where(np.all(res == b, axis=1), w, samples)
-
-            return samples
-
-        return 1 - 2 * self.state[wires[0]]
+    def apply_parametric_program(self, operations, **kwargs):
+        # pylint: disable=attribute-defined-outside-init
+        rotations = kwargs.get("rotations", [])
+
+        # Storing the active wires
+        self._active_wires = ForestDevice.active_wires(operations + rotations)
+
+        # Apply the circuit operations
+        for i, operation in enumerate(operations):
+            # number of wires on device
+            wires = self.remap_wires(operation.wires)
+
+            if i > 0 and operation.name in ("QubitStateVector", "BasisState"):
+                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
+                                  "on a {} device.".format(operation.name, self.short_name))
+
+            # Prepare for parametric compilation
+            par = []
+            for param in operation.params:
+                if isinstance(param, Variable):
+                    # Using the idx for each Variable instance
+                    parameter_string = "theta" + str(param.idx)
+                    if parameter_string not in self._parameter_map:
+                        current_ref = self.prog.declare(parameter_string, "REAL")
+                        self._parameter_reference_map[parameter_string] = current_ref
```

Comment:
It might be good to have line comments here, but if my understanding is correct:

1. We check if the parameter string has been used before(?)

2. If not, a Quil parameter is created (with type real), and added to the parameter reference map

3. Finally, the parameter map maps the parameter string to the parameter value

Reply:
Yes, this flow of logic is good! Added some comments, let me know if it's still unclear